### DISCUSSION TITLE
TreeView is very slow

### DISCUSSION THREAD

**Discussion by D0ctorWh0** [AvaloniaUI/Avalonia#6365] [Q&A] [answered]
I need to display expanded big tree in TreeView, it can contain more then 10000 items. But TreeView starts acting very slow at 1000 items at top level. Is there any way to speed it up?

**Comment by grokys** [comment, marked as answer]
We don't have virtualization on our `TreeView` and we're unlikely to add it, because the way `TreeView` works (nested `ItemsControls`) makes virtualization fraught with problems (and WPF's virtualization has problems aplenty).

In fact if you look in the Visual Studio source code (using ILSpy or similar) they don't use WPF's `TreeView` due to its problems with virtualization.

UWP's tree view uses a much more sensible approach of flattening the tree into a plain list. If you're wanting to support large tree views the best way you can do this currently is:

- Flatten your tree into a list of e.g. `NodeViewModel`s
- Each `NodeViewModel` has an `Indent` property indicating how many levels its indented
- Display the list using a standard `ListBox`
- Use a converter to convert the `Indent` into a margin for each element
- Add an expander button to each list item which when clicked, removes or adds its children to the flat list

It's not ideal I know, and we're talking about creating a control which should support this, but no ETA currently.

**Reply by D0ctorWh0** [reply]
Thanks! I will use this approach, especially since the data for my tree is already presented as a list with nodeID/parentID.

**Comment by kekekeks** [comment]
https://github.com/kekekeks/example-avalonia-huge-tree - an example of ItemsRepeater-based TreeView that can present 100K+ elements.

**Reply by D0ctorWh0** [reply]
Thanks! I searched in google, but never found this example.

**Comment by maxkatz6** [comment]
Also there is an another option now
https://github.com/AvaloniaUI/Avalonia.Controls.TreeDataGrid
